### DISCUSSION TITLE
How to build custom components based on `<Box>`?

### DISCUSSION THREAD

**Discussion by gregoralbrecht** [system-ui/theme-ui#1418] [Q&A] [answered]
Hey everyone 👋  First of all, thank you all for the effort you're putting into this project. It's helped me immensely and even after having read a lot of PRs, issues, and discussions on here, I'm still not grasping the bigger picture. So kudos!

Now, I have read about the [`useVariants()` idea here](https://github.com/system-ui/theme-ui/issues/1247), the confusion about all different [`ThemeProviders`](https://github.com/system-ui/theme-ui/issues/1230) and several comments on TypeScript support. However, I'm still not sure I fully understand the direction in which the upcoming changes regarding `ThemeProvider`s and variants. What I would love to do is: Build custom components similar to the ones in `@theme-ui/components`:

```jsx
import { Box } from '@theme-ui/components'

const Text = React.forwardRef(function Text(props, ref) {
  return (
    <Box
      ref={ref}
      as="p"
      sx={{
        // custom styles
      }}
    />
  );
});
```

I tried a lot of different versions of `@theme-ui/core`, `@theme-ui/components`—and now even `@theme-ui/theme-provider`—but I just couldn't get it to work. Here were my approaches:

##  1. `@latest` versions (`v0.3.5`)
I installed `@theme-ui/core@latest` and `@theme-ui/components@latest`. The `<Box>` and the `sx` prop worked just fine by itself. However, due to TypeScript errors, I was not able to **a) pass a `ForwardRef` to it (like above)** and **b) pass any other props to `<Box>`**. 

## 2. `@next` versions (`v0.6.0-alpha.4`)
I thought—since you were working on types lately—that maybe upgrading to `0.6.0-alpha.4` for both packages would fix the TypeScript errors. It didn't fix it and now the `sx` prop wasn't resolving any values from my theme, even on the `<Box>` coming straight from `@theme-ui/components`.

## 3. Using `theme-ui`
Then, I tried to use `theme-ui` (I tried to avoid it because it includes a lot of things I don't need), but same result as before: Theme values were not resolved. 

# Target API
I would love to achieve something like this:
```jsx
<Text size={2} tone="positive">Hola</Text>
```

I'm coming from `styled-system`, so I'm a bit biased here. But I loved the `variants` API there, which let me do this:
```jsx
const Text = styled('p', { shouldForwardProp })(
  variant({
    prop: 'size',
    variants: {
      tiny: {
        fontSize: 0,
      },
      small: {
        fontSize: 1,
      },
      standard: {
        fontSize: 2,
      },
      large: {
        fontSize: 3,
      },
    },
  }),
  variant({
    prop: 'tone',
    variants: {
      neutral: {
        color: 'grey.800',
      },
      positive: {
        color: 'green.700',
      },
      critical: {
        color: 'red.700',
      },
    },
  }),
)
```

Can anyone give me any hints on
1. How I could make the `sx` prop and `Box` work in TypeScript?
2. How to achieve a similar API like the above for custom components with Theme UI?

Any input would be much appreciated 🙏

**Comment by hasparus** [comment]
Happy New Year @gregoralbrecht 

> It didn't fix it and now the sx prop wasn't resolving any values from my theme, even on the <Box> coming straight from @theme-ui/components.

This sounds like its Emotion context version mismatch (new Theme UI depends on Emotion 11), and if it isn't this then it's serious. Do you have a repro?

---

> How I could make the sx prop and Box work in TypeScript?

It works for me? I'll send you a CodeSandbox in a bit, and we'll try to troubleshoot your issue.

> How to achieve a similar API like the above for custom components with Theme UI?

Well, Theme UI doesn't export a styled component factory, and I think it won't. You can build it on top of sx prop and `theme.variants` or use Stitches if you like this API.

**Reply by gregoralbrecht** [reply]
Happy New Year to you, too! And thank you for the fast response.

This is the weirdest thing. Apparently, I messed up big time trying to get this working locally, because in the CodeSandbox it works: https://codesandbox.io/s/text-component-for-gregor-forked-mxefb?file=/src/Box.test.tsx

I wasn't too set on the styled component factory, more on the idea to have the variants defined as you did in your CodeSandbox.

**Comment by hasparus** [comment, marked as answer]
Okay, so one TypeScript error you can see is

```
Type 'RefObject<unknown>' is not assignable to type 'RefObject<HTMLDivElement>'.
```

[**Here's a CodeSandbox for you.**](https://codesandbox.io/s/text-component-for-gregor-0f7h3?file=/src/App.tsx)
Ping me if you have any more problems :)

![image](https://user-images.githubusercontent.com/15332326/103575125-7ac82300-4ed1-11eb-841f-aaa963566fbc.png)
![image](https://user-images.githubusercontent.com/15332326/103575141-83205e00-4ed1-11eb-888d-a2878d785e5e.png)

---

**Sidenote**: `@theme-ui/components` already exports **Text**. If you want to build your own primitive components you can ignore it entirely, and just use `@theme-ui/core`. You can build your own lightweight `Box` component with JSX pragma or manually using `jsx` imported from `@theme-ui/core`.

---

The root cause of the problem is the fact that `React.forwardRef` infers `props: { children?: ReactNode }, ref: Ref<unknown>`, so you assigned `Ref<unknown>` to `Ref<HTMLDivElement>`.

This doesn't work because.
- Ref<T> is covariant in `T` as `RefObject` (the thing you get from useRef, de facto `{ current: T }`)  is covariant, and ref function definition in @types/react is using bivariance hack
- Knowing that `Ref` is covariant, `Ref<X>` is assignable to `Ref<Y>` when `X` is assignable to `Y`
- `unknown` isn't assignable to `HTMLDivElement` 

```ts
    type Ref<T> = { bivarianceHack(instance: T | null): void }["bivarianceHack"] | RefObject<T> | null;
    interface RefObject<T> {
        readonly current: T | null;
    }
```

**Reply by gregoralbrecht** [reply]
Thank you so much! That's great ❤️  

What do you mean with
> we don't support `as` prop on Box 

? [Is this not what is going on here?](https://github.com/system-ui/theme-ui/blob/develop/packages/components/src/Text.js#L6) I'm asking because it's also in your CodesandBox. Or do you mean a different Box?

> You can build your own lightweight Box component

This is exactly what I did in an earlier iteration, but without TypeScript. Due to my lack of TS knowledge, I couldn't get it to work with TS. I was aiming for something similar to what's happening in [@theme-ui/components here](https://github.com/system-ui/theme-ui/blob/develop/packages/components/src/Box.js)

I'm pretty new to TypeScript, so I'm not even going to pretend to understand your last paragraph right now 🙈 Thank you very much for the details, though, will try to understand it!

**Reply by hasparus** [reply]
> Or do you mean a different Box?

Fair point. Sorry, I didn't express it in the best way :/

We don't **fully** support `as` prop in TypeScript, i.e. we don't change types of Box props when we see the `as` prop, so you can't do this in TS
![image](https://user-images.githubusercontent.com/15332326/103582288-d698a900-4edd-11eb-9829-d90362fcb625.png)
despite that it works at runtime.

It's a bit hard on compilation/IDE-suggestion times.
I'd generally advise to use JSX pragma (so you can use `sx` on every component which accepts className) or the components from `@theme-ui/components`  to TypeScript users encountering this problem.

(Edited the previous post so it's not confusing for other people).

> I'm pretty new to TypeScript, so I'm not even going to pretend to understand your last paragraph right now see_no_evil Thank you very much for the details, though, will try to understand it!

I noticed that "PhD candidate" in you bio, so I assumed you'll either react with "it's trivial, in hindsight" or "I don't get it, so exciting".
Sorry if it's too much details or messy :P

Anyway, as an advice for TypeScript newcomer, if you see an `unknown` somewhere in the error message, and you didn't explictly use `unknown`, try hovering over variables/parameters that you didn't annotate, and see if you can narrow that `unknown` either by annotating parameters or explicitly passing types to a generic function.

**Reply by hasparus** [reply]
BTW thanks so much for posting this in Discussions and stating it as a question instead of writing an issue titled "Box doesn't work in TypeScript" :smile:

**Reply by gregoralbrecht** [reply]
Thanks for the advice! It's definitely still _"I don't get it, so exciting"_ for me right now, so I will read more about this! The degree is not computer science, though, so I will have to freshen up my memory on bivariance... And yeah, I figured Discussions was more appropriate since I was clearly just not using it correctly—which was proven by your CodeSandbox. Thanks again!
